builtin/logical/database/version_wrapper.go

```diff
@@ -0,0 +1,268 @@
+package database
+
+import (
+	"context"
+	"crypto/rand"
+	"fmt"
+
+	log "github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/vault/helper/random"
+	"github.com/hashicorp/vault/sdk/database/dbplugin"
+	"github.com/hashicorp/vault/sdk/database/newdbplugin"
+	"github.com/hashicorp/vault/sdk/helper/pluginutil"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+type databaseVersionWrapper struct {
+	v4 dbplugin.Database
+	v5 newdbplugin.Database
+}
+
+// newDatabaseWrapper figures out which version of the database the pluginName is referring to and returns a wrapper object
+// that can be used to make operations on the underlying database plugin.
+func newDatabaseWrapper(ctx context.Context, pluginName string, sys pluginutil.LookRunnerUtil, logger log.Logger) (dbw databaseVersionWrapper, err error) {
+	newDB, err := newdbplugin.PluginFactory(ctx, pluginName, sys, logger)
+	if err == nil {
+		dbw = databaseVersionWrapper{
+			v5: newDB,
+		}
+		return dbw, nil
+	}
+
+	legacyDB, err := dbplugin.PluginFactory(ctx, pluginName, sys, logger)
+	if err == nil {
+		dbw = databaseVersionWrapper{
+			v4: legacyDB,
+		}
+		return dbw, nil
+	}
+
+	return dbw, fmt.Errorf("invalid database version")
+}
+
+// Initialize the underlying database. This is analogous to a constructor on the database plugin object.
+// Errors if the wrapper does not contain an underlying database.
+func (d databaseVersionWrapper) Initialize(ctx context.Context, req newdbplugin.InitializeRequest) (newdbplugin.InitializeResponse, error) {
+	if !d.isNewDB() && !d.isLegacyDB() {
+		return newdbplugin.InitializeResponse{}, fmt.Errorf("no underlying database specified")
+	}
+
+	// v5 Database
+	if d.isNewDB() {
+		return d.v5.Initialize(ctx, req)
+	}
+
+	// v4 Database
+	saveConfig, err := d.v4.Init(ctx, req.Config, req.VerifyConnection)
+	if err != nil {
+		return newdbplugin.InitializeResponse{}, err
+	}
+	resp := newdbplugin.InitializeResponse{
+		Config: saveConfig,
+	}
+	return resp, nil
+}
+
+// NewUser in the database. This is different from the v5 Database in that it returns a password as well.
+// This is done because the v4 Database is expected to generate a password and return it. The NewUserReponse
+// does not have a way of returning the password so this function signature needs to be different.
+// The password returned here should be considered the source of truth, not the provided password.
+// Errors if the wrapper does not contain an underlying database.
+func (d databaseVersionWrapper) NewUser(ctx context.Context, req newdbplugin.NewUserRequest) (resp newdbplugin.NewUserResponse, password string, err error) {
```

Comment:
This probably okay as-is, but I also wanted to bring up a suggestion that was tossed around, which is the idea of having the password be part of `newdbplugin.NewUserResponse` (and its proto equivalent), but marking that field as being there solely for backward compatibility purposes. This should address the issue of `NewUser` being unable to return the password string if we had a wrapper around v4. We could probably do something in the same vein (e.g. an `UpdateType` or similar on the request payload) for `UpdateUser` so that the v4 wrapper knows what method to call.

Reply:
I don't think this is a good idea for three reasons:
1) It conflates the two interfaces. We'd be stuck with the new interface returning a password even if we removed the old interface.
2) It confuses plugin authors. `NewUser` is receiving a password and returning one?
3) We're explicitly moving password generation out of plugins and into Vault. By having the response return a password, we revert that decision.

Unfortunately, we cannot make the version selection code simpler because we are dramatically changing what parts of the system are responsible for what (in particular password generation).

Reply:
That's right, I'm satisfied with the current approach since there are quite a few downsides for little gain going with the other suggestion. Documentation around fields should hopefully help alleviate with the confusion on the return value. However, leaking some of the old behavior that will be unused in v5 and having dangling fields when v4 is removed is certainly non-ideal.